File: Algorithms/Hash Table/36. Valid Sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
		D = set()
		for inxR, row in enumerate(board):
			for inxL, num in enumerate(row):
				if(num != '.'):
					if('{}({})'.format(inxL, num) in D or '({}){}'.format(num, inxR) in D or '{}({}){}'.format(inxL // 3, num, inxR // 3) in D):
						return False
					D.add('{}({})'.format(inxL, num))
					D.add('{}({}){}'.format(inxL // 3, num, inxR // 3))
					D.add('({}){}'.format(num, inxR))
		return True

class Solution(object):
    def isValidSudoku(self, board):
		D = set()
		for inxR, row in enumerate(board):
			for inxL, num in enumerate(row):
				if(num != '.'):
					if((inxL, num) in D or (num, inxR) in D or (inxL // 3, num, inxR // 3) in D):
						return False
					D.add((inxL, num))
					D.add((inxL // 3, num, inxR // 3))
					D.add((num, inxR))
		return True

File: Algorithms/Hash Table/36. Valid Sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        HashSet seen = new HashSet();
        for(int i = 0; i < 9; ++i) {
            for(int j = 0; j < 9; ++j) {
                if(board[i][j] != '.') {
                    String code = "(" + board[i][j] + ")";
                    if(!seen.add(i + code) || !seen.add(code + j) || !seen.add(i/3 + code + j/3)) return false;
                }
            }
        }
        return true;
    }
}
